Lib/_pyrepl/utils.py

```diff
@@ -73,5 +308,43 @@ def disp_str(buffer: str) -> tuple[CharBuffer, CharWidths]:
         else:
             chars.append(c)
             char_widths.append(str_width(c))
-    trace("disp_str({buffer}) = {s}, {b}", buffer=repr(buffer), s=chars, b=char_widths)
+
+        if colors and colors[0].span.end == i:  # current color ends now
+            post_color = _colorize.theme["RESET"]
+            colors.pop(0)
+
+        chars[-1] = pre_color + chars[-1] + post_color
+        pre_color = ""
+        post_color = ""
+
+    if colors and colors[0].span.start < i and colors[0].span.end > i:
```

Comment:
IIUC this reset only occurs at the very last character. If disp_str() is called again with the next line, and the span is still active, it will start without a reset  leading to incorrect coloring beyond the span, no? I am missing anything?

Reply:
`disp_str()` is called every line. For the case where coloring should end with the line end, we would already emit RESET on line 313.

This `if` is an *extra* RESET for multiline strings, because the next line will be starting with a prompt, which has its own coloring.

Then, on the subsequent call to `disp_str()` on line 291 we'll see that we are in the middle of a span and resume the span. (which by the way is exactly why we use a list for the color spans, as entries are consumed one by one but the "currently active one" must still be kept alive)